#145 : Added support for COMMAND COUNT command

VipinRaiP opened the pull request:
### Summary
Added support for  count command. Please review.
COMMAND COUNT is used to get total number of commands in dicedb.
redis documentation : https://redis.io/docs/latest/commands/command-count
This is my first PR and just getting started with the project. I have referred @yashs360's PR #150 for making the code changes. Thank you @yashs360 

### Changes

- Reused used the changes made by @yashs360  in PR #150 for adding constants for command names. Thanks @yashs360.
- Updated the eval methods as per PR #158  to support COMMAND|subcommand, thanks @kakdeykaushik
-  Added method to get the total number of commands.
- Added test and benchmark.

Benchmark results

<img width="927" alt="image" src="https://github.com/user-attachments/assets/b6de961a-074a-461f-afa1-1f9b45e8654f">




Issue : #145

arpitbbhayani commented on tests/command_count_test.go:
```diff
@@ -0,0 +1,41 @@
+func TestCommandCOUNT(t *testing.T) {
+	commandsCount := 40
+
+	subscriber := getLocalConnection()
```
Why call this subscriber? This is just a simple local TCP connection, so name it `connection` at best.

VipinRaiP replied:
Sure

arpitbbhayani commented on tests/command_count_test.go:
```diff
@@ -0,0 +1,41 @@
+		t.Fail()
+	}
+
+	assert.Assert(t, responseCount == int64(commandsCount),
```
instead of typecasting this to `int64` you can set commandCount as `int64` while declaring. Also, is this a good idea to test this? So, every time you add a new command, this test is bound to fail. Can you confirm how is Redis codebase testing this?

VipinRaiP replied:
May i know where can I find the unit tests in redis codebase?

arpitbbhayani commented on tests/command_count_test.go:
```diff
@@ -0,0 +1,41 @@
+		fmt.Sprintf("Unexpected number of CLI commands found. %d expected, %d found", commandsCount, responseCount))
+}
+
+func call(howmany int, t *testing.B) {
```
Giving such generic function name is a really bad idea as this function is available throughout the package. Please reer to other benchmark functions and go's best practices and structure this better.

Also, update the output of number of allocs and time taken during the benchmar.

VipinRaiP replied:
Updated the method name; Followed the benchmark standards of go.
Alloc and time taken will be printed by default when the test runs.

arpitbbhayani commented on tests/command_count_test.go:
```diff
@@ -0,0 +1,41 @@
+	}
+}
+
+func BenchmarkCountCommand200(t *testing.B) {
+	call(200, t)
+}
+
+func BenchmarkCountCommand2000(t *testing.B) {
+	call(2000, t)
+}
```
What is the purpose of passing `200` or `2000` here? Should it just keep firing the `COMMAND COUNT` command, essentially not needing the inner for loop?

Also, the name of the command is `COMMAND COUNT` and not `COUNT`. Please go through the Redis documentation linkedin the issue description.

VipinRaiP replied:
Updated the command name to COMMAND COUNT;

arpitbbhayani commented:
@VipinRaiP in case you missed it. Let me know by where are you planning to wrap it up?

VipinRaiP commented:
> @VipinRaiP in case you missed it. Let me know by where are you planning to wrap it up?

Hi @arpitbbhayani 
Sorry for the delay, i will fix these issues and push updated code over this weekend .
Please excuse

VipinRaiP commented:
> @VipinRaiP in case you missed it. Let me know by where are you planning to wrap it up?

I have fixed the issues and pushed a new commit, thanks

JyotinderSingh commented on tests/command_count_test.go:
```diff
@@ -0,0 +1,34 @@
+}
+
+func getCommandCount(t *testing.B) {
+	connection := getLocalConnection()
```
Is it a good idea to include connection establishment as a part of the benchmark code? Maybe the connection should be created once in the caller and be reused across function calls for the benchmark.

VipinRaiP replied:
I have updated to reuse connection.

JyotinderSingh commented on tests/command_count_test.go:
```diff
@@ -0,0 +1,34 @@
+		fmt.Sprintf("Unexpected number of CLI commands found. expected greater than 0, %d found", responseCount))
+}
+
+func getCommandCount(t *testing.B) {
```
This function can probably be reused for the TestCommandCount method as well.

VipinRaiP replied:
Removed duplicate code and pushed new changes.

JyotinderSingh commented on core/eval.go:
```diff
@@ -21,11 +21,63 @@ var RESP_MINUS_1 []byte = []byte(":-1\r\n")
+	ABORT        = "ABORT"
+)
+
+var diceCommands = []string{
+	PING, SET, GET, TTL, DEL, EXPIRE, BGREWRITEAOF, INCR, INFO, CLIENT, LATENCY, LRU,
+	SLEEP, QINTINS, QINTREM, QINTLEN, QINTPEEK, BFINIT, BFADD, BFEXISTS, BFINFO,
+	QREFINS, QREFREM, QREFLEN, QREFPEEK, STACKINTPUSH, STACKINTPOP, STACKINTLEN,
+	STACKINTPEEK, STACKREFPUSH, STACKREFPOP, STACKREFLEN, STACKREFPEEK, SUBSCRIBE,
+	QWATCH, COMMAND, MULTI, EXEC, DISCARD, ABORT}
```
Let's instead create a fixed-size array instead of a slice since it will never grow at runtime.

VipinRaiP replied:
sure

JyotinderSingh commented on core/eval.go:
```diff
@@ -863,95 +915,117 @@ func evalQWATCH(args []string, c *Client) []byte {
+
+// evalCOMMANDCOUNT returns an number of commands supported by DiceDB
+func evalCOMMANDCOUNT() []byte {
+	return Encode(len(diceCommands), false)
```
instead of calculating the length of the commands for every request, we can simply store it in a constant and return that. We can add a runtime check to verify that the constant's value does not get outdated.
```go
const (
    diceCommandsCount = len(diceCommands)
)

func init() {
    if len(diceCommands) != diceCommandsCount {
        panic("diceCommands slice has been modified")
    }
}
```

JyotinderSingh commented on core/eval.go:
```diff
@@ -863,95 +915,117 @@ func evalQWATCH(args []string, c *Client) []byte {
+// COUNT: return total count of commands in Dice.
+func evalCOMMAND(args []string) []byte {
+	if len(args) == 0 {
+		return Encode(fmt.Errorf("(error) ERR wrong number of arguments for 'command' command"), false)
```
nit: we could use errors.New() instead of fmt.Errorf().

There's not much difference, but since there are no custom args for this string we probably don't need fmt.Errorf

JyotinderSingh commented:
Left some minor comments.

gauravsarma1992 commented:
This PR https://github.com/DiceDB/dice/pull/170 will have some conflicts with the current PR.

VipinRaiP commented:
> This PR #170 will have some conflicts with the current PR.

Should this PR wait until #170  is merged and make changes as per new code?

JyotinderSingh commented:
> > This PR #170 will have some conflicts with the current PR.
> 
> Should this PR wait until #170 is merged and make changes as per new code?

Hi @VipinRaiP, thanks for all the changes so far. #170 has been merged now. Please rebase your changes on top of the latest master and fix any conflicts that may be there. Once you have addressed the reviews, please resolve those comments and I will review this again.

VipinRaiP commented:
Sure

VipinRaiP commented:
> > > This PR #170 will have some conflicts with the current PR.
> > 
> > 
> > Should this PR wait until #170 is merged and make changes as per new code?
> 
> Hi @VipinRaiP, thanks for all the changes so far. #170 has been merged now. Please rebase your changes on top of the latest master and fix any conflicts that may be there. Once you have addressed the reviews, please resolve those comments and I will review this again.

Rebased and pushed a new commit, please review

JyotinderSingh commented:
Approved, will merge soon.
Could you please update the benchmark screenshot (since we have made some changes now) @VipinRaiP

JyotinderSingh commented:
@VipinRaiP Looks like a unit test is failing with these changes, could you please look into it?
